stdlib/public/core/Collection.swift

```diff
@@ -260,9 +260,10 @@ extension CollectionType {
 
   /// Returns the first element of `self`, or `nil` if `self` is empty.
   ///
-  /// - Complexity: O(1)
+  /// - Complexity: O(1) in most cases, possibly O(N) for lazy collections.
   public var first: Generator.Element? {
-    return isEmpty ? nil : self[startIndex]
+    var gen = generate()
+    return gen.next()
```

Comment:
I'm also very mildly concerned that, despite `endIndex` being documented as O(1), it's possible for some collection to implement it more expensively (although they really shouldn't do that). Conversely, it's possible for a collection to implement a generator such that the creation of the generator is expensive, but they shouldn't do that either. Since iteration of sequences (and therefore generators) is so prevalent and therefore a prime target for optimization, I'm assuming that the generator approach (as used in my PR) is more reliably efficient than the index approach.